daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/LimitsSpec.scala

```diff
@@ -0,0 +1,436 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package speedy
+
+import com.daml.lf.data.{FrontStack, ImmArray, Ref, Time}
+import com.daml.lf.interpretation.{Error => IE}
+import com.daml.lf.language.{Ast, PackageInterface}
+import com.daml.lf.testing.parser.Implicits._
+import com.daml.lf.transaction.{SubmittedTransaction, TransactionVersion, Versioned}
+import com.daml.lf.validation.Validation
+import com.daml.lf.value.Value
+import org.scalatest.Inside
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class LimitsSpec extends AnyWordSpec with Matchers with Inside {
+
+  private[this] val pkgs = typeAndCompile(
+    p"""
+      module Mod {
+        
+        record @serializable T = { 
+          signatories: List Party,
+          observers: List Party
+        };
+        
+        record @serializable NoOpArg = { controllers: List Party, observers: List Party };
+        
+        val fetches: List (ContractId Mod:T) -> Update (List Mod:T) = 
+          \(cids: List (ContractId Mod:T)) -> 
+            case cids of 
+              Cons h t -> 
+                ubind 
+                  first: Mod:T <- fetch @Mod:T h;
+                  rest: List Mod:T <- Mod:fetches t
+                in 
+                  upure @(List Mod:T) Cons @Mod:T [first] rest
+            | Nil -> 
+                upure @(List Mod:T) Nil @Mod:T;
+        
+        template (this : T) =  {
+          precondition True;
+          signatories Mod:T {signatories} this;
+          observers Mod:T {observers} this;
+          agreement "Agreement";
+          choice @nonConsuming NoOp (self) (arg: Mod:NoOpArg): Unit,
+            controllers Mod:NoOpArg {controllers} arg,
+            observers Mod:NoOpArg {observers} arg
+            to
+              upure @Unit ();
+         };
+      }
+    """
+  )
+
+  "Machine" should {
+
+    val committers = (0 to 99).view.map(i => Ref.Party.assertFromString(s"Parties$i")).toSet
+
+    "refuse to build contract with too many signatories" in {
+      val limit = 10
+      val limits = interpretation.Limits.Lenient.copy(contractSignatories = limit)
+      val e =
+        e"""\(signatories: List Party) (observers: List Party) -> 
+             create @Mod:T Mod:T { signatories = signatories, observers = observers }
+         """
+
+      List(1, limit, limit + 1, 99).count { i =>
```

Comment:
Have you considered to instead test which ones succeed and which ones fail instead of only testing how many fail? That might make the test easier to understand and makes sure we’re testing what we want to test.

Reply:
it also seems like decent parts of this logic are replicated. Maybe worth considering if we can factor it out but I don’t feel strongly about that.

Reply:
Surprising difficult to factorize.

Reply:
067df150f7dfc933ed471da32391bf176d5fd7b2